storage/src/linear/io_uring.rs

```diff
@@ -0,0 +1,811 @@
+                batch.outstanding.len(),
+                batch.backlog.len(),
+                batch.sq.len(),
+                batch.sq.is_full(),
+                batch.cq.len(),
+                batch.cq.is_full(),
+                batch.entries.size_hint(),
+            );
+
+            // Top of the loop: synchronize the completion queue pointers and
+            // maybe block until at least one completion is available.
+            if let Err(err) = batch.sync_completion_queue() {
+                bail!(err);
+            }
+
+            // Drain the completion queue and backlog any entries we need to
+            // re-submit before flushing to the submission queue.
+            batch.flush_completion_queue(&mut errors);
+
+            // Finally, try to flush more entries to the submission queue
+            // prioritizing resubmits before new entries.
+            if let Err(err) = batch.flush_submission_queue() {
+                bail!(err);
+            }
+        }
+
+        // batch.is_finished() is true here, so we can forget the drop guard
+        let outstanding = batch.outstanding.forget();
+        debug_assert!(
+            outstanding.is_empty(),
+            "io-uring write batch finished with {} outstanding entries",
+            outstanding.len()
+        );
+
+        if errors.is_empty() {
+            Ok(batch.written)
+        } else {
+            Err(BatchErrors { errors })
+        }
+    }
+}
+
+impl<'a> QueueEntry<'a> {
+    fn new((offset, buffer): (u64, &'a [u8])) -> Self {
+        // trivial check because we never allocate nodes larger than 16 MiB (AREA_SIZES)
+        debug_assert!(u32::try_from(buffer.len()).is_ok());
+        Self {
+            original_offset: offset,
+            offset,
+            pointer: buffer.as_ptr(),
+            length: buffer.len() as u32,
+            lifetime: std::marker::PhantomData,
+        }
+    }
+
+    fn build_submission_queue_entry(&self, fd: RawFd) -> io_uring::squeue::Entry {
+        io_uring::opcode::Write::new(io_uring::types::Fd(fd), self.pointer, self.length)
+            .offset(self.offset)
+            .build()
+            .user_data(self.original_offset)
+    }
+
+    /// Handles the result of a completed write operation.
+    ///
+    /// If the operation was successful, returns the number of bytes written. If
+    /// the operation failed, returns a `BatchError` describing the failure.
+    ///
+    /// If `length` is non-zero after a successful write, the entry must be
+    /// re-submitted to write the remaining data.
+    fn handle_result(&mut self, res: i32) -> Result<usize, BatchError> {
+        #![expect(clippy::arithmetic_side_effects, clippy::cast_sign_loss)]
+
+        if res < 0 {
+            let err = std::io::Error::from_raw_os_error(-res);
+            // rust stdlib maps EAGIN to WouldBlock; re-submit in that case
```

Comment:
```suggestion
            // rust stdlib maps EAGAIN to WouldBlock; re-submit in that case
```

Reply:
`EAGIN` is the error code, not `EAGAIN`.